pygeo/DVGeometry.py

```diff
@@ -2133,6 +2133,81 @@ def writePlot3d(self, fileName):
             """
         self.FFD.writePlot3dCoef(fileName)
 
+    def updatePyGeo(self, geo, outputType, fileName, nRefU=0, nRefV=0):
+        """ Deform a pyGeo object and write to a file of specified type
+        given the (deformed) current state of the FFD object.
+
+        Parameters
+        ----------
+        geo : pyGeo object
+            A pyGeo object containing an initialized object
+        outputType: str
+            Type of output file to be written. Can be `iges` or `tecplot`
+        fileName: str
+            Filename for the output file. Should have no extension, an
+            extension will be added
+        nRefU: int or list of ints
+            Number of spline refinement points to add in the surface B-Spline u-direction.
+            If scalar, it is applied across each surface. If list, the length must match the
+            number of surfaces in the object and corresponding entries are matched with surfaces.
+        nRefV: int or list of ints
+            Number of spline refinement points to add in the surface B-Spline v-direction.
+            If scalar, it is applied across each surface. If list, the length must match the
+            number of surfaces in the object and corresponding entries are matched with surfaces
+        """
+        # Refine Surface -- U-Direction
+        if isinstance(nRefU, int):
+            # Refine BSplines by adding knot points
+            Refine_U = numpy.linspace(0.0, 1.0, nRefU)
```

Comment:
I am not sure how knots are added in pygeo. What would happen if one of the u locations correspond to an existing knot? I think if there is no check for this, it will change the continuity of the bspline at that knot. Do we check this? I am not necessarily asking for changes here, just trying to understand.

Also, my bspline math is a bit rusty, but I think the u=0 and 1 positions will have as many knots as the order of the bspline (or something similar?). Does adding more points by default make sense there? Maybe we can just add 2 to `nRefU`, get the linspace, and then dont add the knots at u=0 and 1 locations.

Reply:
For your second point, knot points are automatically *not* added at the ends of the spline where `u = 0 or 1` (there is a check for this in pySpline [here](https://github.com/mdolab/pyspline/blob/a3b45e9aad5938b09a318b5f570bd0335dabfc55/pyspline/pySpline.py#L1795)). You bring up a good point that `nRefU = 0 or 1` cause no refinement (for different reasons), so I will offset that to refine properly.

For your first point, I am actually not sure either. From what I can tell from the [`insertKnot.f90`](https://github.com/mdolab/pyspline/blob/master/src/insertKnot.f90) file, there is a check for multiplicity:

```
! Compute its multiplicity
s = 0 ! Knot multiplicity
mult: do i=0,k-1
   if (abs(t(ileft-i) -u ) < 1e-12) then 
      s = s + 1
   else
      exit mult
   end if
end do mult

! We need to make sure that the requested multipliity r, plus 
! the actual multiplicity of this know is less than (k-1)

if (s + r + 1 > k) then
   r = k-1-s
end if

! If we *can't* insert the knot, we MUST copy t and coef to t_new
! and coef_new and return
if (r == 0) then
   coef_new(:,1:nctl) = coef(:,1:nctl)
   t_new(1:nctl+k) = t(1:nctl+k)
   return
end if
```

My knowledge of bspline math is limited, so I am not entirely sure what the process here is. But, it seems that if there is multiplicity and it will cause the number of knot points to exceed the order at that point, the inserted point is simply not added. This may cause unexpected behavior for the user if their geometry is refined less than expected, but realistically if their geometry already has a point where it is being refined it should not make a big difference. I can add a warning about this in the docs.

Is this an issue we want to open for pySpline?

Reply:
I see, I think pyspline is fine, but we should _not_ add the knot if there is another knot there. Changing the multiplicity from 1 to 2 would change the continuity of the derivatives of the bspline, and instead of refining it, it will change the curve itself. Is there an easy way we can get the current knot vector to check we are only refining and not modifying the multiplicity anywhere?

Reply:
I will take a look to see if this is possible.

Reply:
Sounds great. The addition of 2 makes sense if pyspline is not adding the knots at 0 and 1 by default. Let me know if this is possible, if it takes too much work, you can also just add a few comments explaining this behavior so that people are forced to understand what the bspline knots are doing? 

Anyways, besides this the PR looks good for me.

Reply:
It is hard coded. I'll hard code it here too.

Reply:
Yeah then hard coding is just fine. Leave a comment in the code that either points here or explains why 1e-12 is there, and we should be set. (Out of curiosity, how does the knot vector look like for your example? Does it actually have values that might be overlapping or was my suggestion a complete non-issue?)

Reply:
I updated it and added an internal function to check the tolerance. Let me know if it makes sense.

Reply:
Neat, the check looks great. I am not sure how big of a difference you would get; it depends on the order of the spline and the actual positions of the control points etc etc but I think doing it this way should be more failsafe. The rest of the PR looks good to me!